cmd/tools/cloudburst/cloudburst.go

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	cloud "github.com/mattermost/mattermost-cloud/model"
+	mmv1alpha1 "github.com/mattermost/mattermost-operator/apis/mattermost/v1alpha1"
+	log "github.com/sirupsen/logrus"
+)
+
+var logger *log.Logger
+
+type completedReport struct {
+	installation *cloud.Installation
+	completedAt  time.Time
+	createdAt    time.Time
+}
+
+type errorReport struct {
+	installation *cloud.Installation
+	timestamp    time.Time
+	message      string
+}
+
+// ReportType identifies types of test reports that may be returned
+type ReportType string
+
+const (
+	// ErrorReportType represents an error produced by a test
+	ErrorReportType ReportType = "errorReport"
+	// CompletedReportType represents a result from a completed test
+	CompletedReportType ReportType = "completedReport"
+)
+
+// Report allows for both error reports and completion reports to be
+// passed along the same channel with type safety
+type Report interface {
+	// Type returns the ReportType held by the concrete implementation
+	// of Report
+	Type() ReportType
+}
+
+// Type returns ErrorReportType for errorReport objects
+func (e *errorReport) Type() ReportType {
+	return ErrorReportType
+}
+
+// Type returns CompletedReportType for completedReport objects
+func (c *completedReport) Type() ReportType {
+	return CompletedReportType
+}
+
+// Blaster represents a single instance of the load testing tool. Each
+// Blaster is aware of its own reports (errors and completed reports),
+// and contains internal state that pertains to the entire test run.
+type Blaster struct {
+	client  *cloud.Client
+	testID  string
+	group   *cloud.Group
+	reports []Report
+}
+
+// NewBlaster creates a new Blaster with which to blast the
+// Provisioning API
+func NewBlaster(serverAddress string) *Blaster {
+	client := cloud.NewClient(serverAddress)
+	return &Blaster{client: client, testID: cloud.NewID()}
+}
+
+var blastCommand = &cobra.Command{
+	Use:   "cloudburst",
+	Short: "Run a load test against a provisioning server.",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		serverAddress, _ := cmd.Flags().GetString("server")
+		batchSize, _ := cmd.Flags().GetInt("batch")
+		total, _ := cmd.Flags().GetInt("total")
+		runs, _ := cmd.Flags().GetInt("runs")
+
+		logger.SetLevel(log.DebugLevel)
+		logger.Infof("Server address %s", serverAddress)
+		logger.WithField("server", serverAddress)
+		blaster := NewBlaster(serverAddress)
+		err := blaster.createGroup()
+		if err != nil {
+			return errors.Wrap(err, "failed to create group for test installations")
+		}
+
+		logger.Infof("Waiting for Group to be created..")
+		blaster.waitForGroup()
```

Comment:
Isn't the group creation synchronous operation? I am not sure if we need to wait for a group to be created.

Reply:
It is not synchronous however the aggressive retrial strategy in the routines that send `CreateInstallationRequest`s does probably make this redundant. I will double-check and remove this if I find that it is not necessary.